runtime/tests/interpreter/values_test.go

```diff
@@ -0,0 +1,1565 @@
+)
+
+// TODO: make these program args?
+const containerMaxDepth = 3
+const containerMaxSize = 1_000
+const innerContainerMaxSize = 300
+const compositeMaxFields = 10
+
+var runSmokeTests = flag.Bool("runSmokeTests", false, "Run smoke tests on values")
+
+func TestRandomMapOperations(t *testing.T) {
+	if !*runSmokeTests {
+		t.SkipNow()
+	}
+
+	seed := time.Now().UnixNano()
+	fmt.Printf("Seed used for map opearations test: %d \n", seed)
+	rand.Seed(seed)
+
+	storage := interpreter.NewInMemoryStorage()
+	inter, err := interpreter.NewInterpreter(
+		&interpreter.Program{
+			Program:     ast.NewProgram([]ast.Declaration{}),
+			Elaboration: sema.NewElaboration(),
+		},
+		utils.TestLocation,
+		interpreter.WithStorage(storage),
+		interpreter.WithImportLocationHandler(
+			func(inter *interpreter.Interpreter, location common.Location) interpreter.Import {
+				return interpreter.VirtualImport{
+					Elaboration: inter.Program.Elaboration,
+				}
+			},
+		),
+	)
+	require.NoError(t, err)
+
+	numberOfValues := randomInt(containerMaxSize)
+
+	var testMap, copyOfTestMap *interpreter.DictionaryValue
+	var storageSize, slabCounts int
+
+	entries := newValueMap(numberOfValues)
+	orgOwner := common.Address{'A'}
+
+	t.Run("construction", func(t *testing.T) {
+		keyValues := make([]interpreter.Value, numberOfValues*2)
+		for i := 0; i < numberOfValues; i++ {
+			key := randomHashableValue(inter)
+			value := randomStorableValue(inter, 0)
+
+			entries.put(inter, key, value)
+
+			keyValues[i*2] = key
+			keyValues[i*2+1] = value
+		}
+
+		testMap = interpreter.NewDictionaryValueWithAddress(
+			inter,
+			interpreter.DictionaryStaticType{
+				KeyType:   interpreter.PrimitiveStaticTypeAnyStruct,
+				ValueType: interpreter.PrimitiveStaticTypeAnyStruct,
+			},
+			orgOwner,
+			keyValues...,
+		)
+
+		storageSize, slabCounts = getSlabStorageSize(t, storage)
+
+		require.Equal(t, testMap.Count(), entries.size())
+
+		entries.foreach(func(orgKey, orgValue interpreter.Value) (exit bool) {
+			exists := testMap.ContainsKey(inter, interpreter.ReturnEmptyLocationRange, orgKey)
+			require.True(t, bool(exists))
+
+			value, found := testMap.Get(inter, interpreter.ReturnEmptyLocationRange, orgKey)
+			require.True(t, found)
+			utils.AssertValuesEqual(t, inter, orgValue, value)
+
+			return false
+		})
+
+		owner := testMap.GetOwner()
+		assert.Equal(t, orgOwner, owner)
+	})
+
+	t.Run("iterate", func(t *testing.T) {
+		require.Equal(t, testMap.Count(), entries.size())
+
+		testMap.Iterate(func(key, value interpreter.Value) (resume bool) {
+			orgValue, ok := entries.get(key)
+			require.True(t, ok, "cannot fine key: %v", key)
+
+			utils.AssertValuesEqual(t, inter, orgValue, value)
+			return true
+		})
+	})
+
+	t.Run("deep copy", func(t *testing.T) {
+		newOwner := atree.Address{'B'}
+		copyOfTestMap = testMap.Transfer(
+			inter,
+			interpreter.ReturnEmptyLocationRange,
+			newOwner,
+			false,
+			nil,
+		).(*interpreter.DictionaryValue)
```

Comment:
Do we need to smoke test `Transfer()` with `remove` flag set to `true`?  Same question for `ArrayValue` and `CompositeValue`.

Reply:
Added in a6d8aa1a4a09979404f17a4c8e103b461be4075a